base/abstractdict.jl

```diff
@@ -502,108 +502,140 @@ push!(t::AbstractDict, p::Pair) = setindex!(t, p.second, p.first)
 push!(t::AbstractDict, p::Pair, q::Pair) = push!(push!(t, p), q)
 push!(t::AbstractDict, p::Pair, q::Pair, r::Pair...) = push!(push!(push!(t, p), q), r...)
 
-# hashing objects by identity
+
+# Hashing objects by identity
+_tablesz(x::Integer) = x < 16 ? 16 : one(x)<<((sizeof(x)<<3)-leading_zeros(x-1))
 
 """
-    ObjectIdDict([itr])
+    IdDict([itr])
 
-`ObjectIdDict()` constructs a hash table where the keys are (always)
-object identities.  Unlike `Dict` it is not parameterized on its key
-and value type and thus its `eltype` is always `Pair{Any,Any}`.
+`IdDict{K,V}()` constructs a hash table using object-id as hash and
+`===` as equality with keys of type `K` and values of type `V`.
 
 See [`Dict`](@ref) for further help.
 """
-mutable struct ObjectIdDict <: AbstractDict{Any,Any}
+mutable struct IdDict{K,V} <: AbstractDict{K,V}
     ht::Vector{Any}
     ndel::Int
-    ObjectIdDict() = new(Vector{Any}(uninitialized, 32), 0)
+    IdDict{K,V}() where {K, V} = new{K,V}(Vector{Any}(uninitialized, 32), 0)
 
-    function ObjectIdDict(itr)
-        d = ObjectIdDict()
+    function IdDict{K,V}(itr) where {K, V}
+        d = IdDict{K,V}()
         for (k,v) in itr; d[k] = v; end
         d
     end
 
-    function ObjectIdDict(pairs::Pair...)
-        d = ObjectIdDict()
+    function IdDict{K,V}(pairs::Pair...) where {K, V}
+        d = IdDict{K,V}()
+        sizehint!(d, length(pairs))
         for (k,v) in pairs; d[k] = v; end
         d
     end
 
-    ObjectIdDict(o::ObjectIdDict) = new(copy(o.ht))
+    IdDict{K,V}(d::IdDict{K,V}) where {K, V} = new{K,V}(copy(d.ht))
+    function IdDict{K,V}(d::IdDict{K}) where {K, V}
+        d = IdDict{K,V}()
+        sizehint!(d, length(pairs))
```

Comment:
This method seems wrong. Is it a mistaken copy-paste of the second method above?

Reply:
Yes, thanks, the `pairs` is wrong.  But the method itself is needed:

My thinking here was to provide a method to allow something like:
```
di = IdDict{Symbol,Int}(:A=>1)
IdDict{Symbol,Float64}(di)
```
the loop doing the `setindex!` will call convert on the integer values of `di`.

I'll fix it and add a test.

Reply:
Actually, it is not needed after all as it will call into the general `function IdDict{K,V}(itr) ...` method.  Thanks!